src/vos/vos_dtx.c

```diff
@@ -717,6 +730,469 @@ vos_dtx_abort_one(struct vos_container *cont, struct daos_tx_id *dti,
 	return rc;
 }
 
+struct vos_dtx_share {
+	/** Link into the daos_tx_handle::dth_shares */
+	d_list_t		vds_link;
+	/** The DTX record type, see enum vos_dtx_record_types. */
+	uint32_t		vds_type;
+	/** The record in the related tree in SCM. */
+	umem_id_t		vds_record;
+};
+
+static bool
+vos_dtx_is_normal_entry(struct umem_instance *umm, umem_id_t entry)
+{
+	if (UMMID_IS_NULL(entry) ||
+	    umem_id_equal(umm, entry, DTX_UMMID_ABORTED) ||
+	    umem_id_equal(umm, entry, DTX_UMMID_UNKNOWN))
+		return false;
+
+	return true;
+}
+
+/* The caller has started PMDK transaction. */
+int
+vos_dtx_register_record(struct umem_instance *umm, umem_id_t record,
+			uint32_t type, uint32_t flags)
+{
+	struct daos_tx_handle		*dth = vos_dth_get();
+	struct vos_dtx_entry_df		*dtx;
+	struct vos_dtx_record_df	*rec;
+	struct vos_dtx_share		*vds;
+	struct vos_dtx_share		*next;
+	umem_id_t			 rec_mmid = UMMID_NULL;
+	umem_id_t			*entry = NULL;
+	uint32_t			*shares = NULL;
+	int				 rc = 0;
+	bool				 shared = false;
+
+	switch (type) {
+	case DTX_RT_OBJ: {
+		struct vos_obj_df	*obj;
+
+		obj = umem_id2ptr(umm, record);
+		entry = &obj->vo_dtx;
+		if (dth == NULL || dth->dth_intent == DAOS_INTENT_UPDATE)
+			shares = &obj->vo_dtx_shares;
```

Comment:
> In my opinion, register & degeister is corresponding to DTX transaction start and commit/abort, so it's more clear to call them when start or commit/abort DTX transaction.

DTX register is similar as umem_tx_add_ptr() for PMDK transaction, it is used as attaching the target (object/key/evt/svt) to be modified to the DTX.

>'vo_dtx_shares' is a transient status used for DTX transaction, it should be in DTX table in my opinion. Storing it in VOS tree introduced metadata unnecessary overhead.

Hm, the 'vo_dtx_shares' is the reference counter for the DTX that shares to update related target (object/key). It is object based. If we move it to DTX table, then it will be DTX record based. But in current model, we do not have efficient way to find out one DTX record with another DTX's ID. So the logic and efficiency will be worse than current patch. I agree with you that current implementation will introduce some overhead, that is trade off between time and space.

Reply:
So I think the question is: Should we put this DTX transaction related information (the DTX uncommitted record is shared by how many DTXs) in persistent VOS tree or in a separate auxiliary index (even in DRAM)? The latter looks better to me.

Reply:
That is not impossible. In fact, related information should be attached to the data record (object/key) in cache, but currently we do not cache key related things. I do not want to introduce too much adjustment about caching keys in DRAM just because of DTX requirement. So be as some temporary solution, it is just put inside object/key on-disk layout. I agree that it will cause more space overhead, we can consider to optimize that next step.